[module-core][Kotlin] Add basic support for sync function

lukmccall opened the pull request:
# Why

Closes ENG-4608.
Adds basic support for sync function.

# How

- Created ExpoModulesHostObject that installs into the runtime as `global.ExpoModules`. A host object is an abstract object where we can define our property getter, allowing us to create JS object for each module lazily.
- Added `isAsync` property to sweet functions and `runSynchronously()` modifier - that will be changed in the future.
- Created `JavaScriptModuleObject` that represents a module in the JS world. It has a map of methods with a cached body.
- All the communication goes via two kotlin methods - `getJavaScriptModuleObject` (which returns a hybrid representation of the module) and `callMethodSync` (which is calling correct implementation). Right now, I'm not using lambdas to pass Kotlin functions into CPP, because I don't see any reason why should I.

# Test Plan

- tested with bare-expo ✅

lukmccall committed:
[module-core][Kotlin] Add basic support for sync function

barthap commented on packages/expo-modules-core/android/build.gradle:
```diff
@@ -108,6 +173,25 @@ dependencies {
+    def rnAAR = rnAARs.singleFile
+    extractJNI(files(rnAAR))
+  } else {
+    println("oooo")
+    // TODO(@lukmccall)
```
fyi if you're curious: related https://github.com/expo/expo/pull/16075#discussion_r838066614

barthap commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/KotlinInteropModuleRegistry.kt:
```diff
@@ -41,26 +41,22 @@ class KotlinInteropModuleRegistry(
-        exportKey(holder.name, methodsInfo)
-        holder.name to methodsInfo
+    return registry.associate { holder ->
+      val methodsInfo = holder.definition.methods.filter { (_, method) -> method.isSync }.map { (name, method) ->
```
- Why do you filter here only sync methods? They weren't filtered before so the asyncs were exported here
- I'd split this, IMO it's difficult to read now:
   ```suggestion
         val methodsInfo = holder.definition.methods
           .filter { (_, method) -> method.isSync }
           .map { (name, method) ->
   ```

lukmccall replied:
It was a mistake - here we want to export only async functions.

barthap commented on packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp:
```diff
@@ -0,0 +1,137 @@
+                                        const jsi::Value &value) {
+  std::string message("RuntimeError: Cannot override the host object for expo module '");
+  message += name.utf8(runtime);
+  message += "'.";
```
I don't think we need a period at the end of the message
```suggestion
  message += "'";
```

barthap commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/jni/JavaScriptModuleObject.kt:
```diff
@@ -0,0 +1,27 @@
+  private external fun initHybrid(): HybridData
+
+  @Suppress("KotlinJniMissingFunction")
+  external fun registerSyncFunction(name: String, args: Int)
```
Looks like we don't register sync functions anywhere, I don't see any call site of this yet. Where is it supposed to be called? 🤔

lukmccall replied:
Yes, you're right. Sorry for that. I did add registration in the module holder.

barthap commented on packages/expo-modules-core/android/src/main/cpp/ExpoModulesHostObject.cpp:
```diff
@@ -0,0 +1,33 @@
+                                const jsi::Value &value) {
+  std::string message("RuntimeError: Cannot override the host object for expo module '");
+  message += name.utf8(runtime);
+  message += "'.";
```
I don't think we need a period at the end of the message
```suggestion
  message += "'";
```

barthap reviewed (commented):
I'm impressed with how you designed the module registration and function call flow.
There are still some parts that I don't understand well, so I left some comments 😉

tsapeta commented on packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp:
```diff
@@ -0,0 +1,137 @@
+
+  return result;
+}
+}
```
I think it's a good practice to add such comments after the closing bracket

```suggestion
} // namespace expo
```

tsapeta commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:
```diff
@@ -43,6 +41,10 @@ class AppContext(
   }
 
+  fun onPostCreate() {
+    jsiInterop.installJSI(reactContextHolder.get()!!.javaScriptContextHolder.get())
```
When could it fail? Can we catch it?

lukmccall replied:
It shouldn't really happen. However, I've replaced it with a little bit safer version.

tsapeta reviewed (commented):
Amazing work! 👏

lukmccall committed:
Apply requested changes

lukmccall committed:
Fix `exportMethods`

lukmccall commented on packages/expo-modules-core/android/src/main/cpp/ExpoModulesHostObject.cpp:
```diff
@@ -0,0 +1,33 @@
+namespace expo {
+
+ExpoModulesHostObject::ExpoModulesHostObject(JSIInteropModuleRegistry *installer) : installer(
+  installer) {}
```
Yes, it does. I'm still fighting with CPP formater inside of the AS. It's so lame 🤷

lukmccall committed:
Add namespace comments

tsapeta commented on packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp:
```diff
@@ -0,0 +1,136 @@
+JavaScriptModuleObject::HostObject::set(jsi::Runtime &runtime, const jsi::PropNameID &name,
+                                        const jsi::Value &value) {
+  std::string message("RuntimeError: Cannot override the host object for expo module '");
+  message += name.utf8(runtime);
```
This line is still necessary (just without the dot) as it closes the single quote 😄 
```suggestion
  message += name.utf8(runtime);
  message += "'";
```

bbarthec reviewed (commented):
I don't think that I would be of any help in this PR 😞
I'm lacking a lot of context and maybe some diagram describing clearly what is what and what is the reason for each cpp/kotlin class 👀

tsapeta commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/functions/AnyFunction.kt:
```diff
@@ -28,7 +46,13 @@ abstract class AnyFunction(
+    throw UnsupportedOperationException("The sync call is not supported yet!")
+  }
+
+
```
```suggestion
```

tsapeta commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:
```diff
@@ -43,6 +41,12 @@ class AppContext(
     }
   }
 
+  fun onPostCreate() {
```
Rename it to something like `installJSIObjects`

wkozyra95 commented on packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp:
```diff
@@ -0,0 +1,136 @@
+
+void JavaScriptModuleObject::registerSyncFunction(jni::alias_ref<jstring> name, jint args) {
+  auto cName = name->toStdString();
+  methodsMetadata.emplace(std::piecewise_construct,
```
that could use some comment :D
also I think we could use c++17, so `methodsMetadata.try_emplace(cName, cName, args, false)` should be equivalent

tsapeta commented on packages/expo-modules-core/android/src/main/cpp/JSIInteropModuleRegistry.cpp:
```diff
@@ -0,0 +1,61 @@
+namespace jni = facebook::jni;
+namespace jsi = facebook::jsi;
+
+namespace expo {
```
Hmm, I think it's ok to just use `expo` for the core things (we also do this on iOS already). It makes sense for GL and AV to have their own namespace as they are separate things and much more specific.

lukmccall committed:
[core][Kotlin] Add cpp part of async functions (#17020)

# Why

Part of https://github.com/expo/expo/pull/16977.
Adds a cpp part of the async function support. 

# How

My implementation is very similar to the TurboModule implementation  - https://github.com/facebook/react-native/blob/main/ReactCommon/react/nativemodule/core/platform/android/ReactCommon/JavaTurboModule.cpp.

- Stored pointer to the js and native callback invoker. The second one is unused right now, but we will need it in the future. 
- Allow modules to register async function on their CPP representation.
- Allow CPP code to call module function with promise. 

# Test Plan

- bare-expo and simple app:
```kotlin
// Cellular module
function("test") { s: String -> s }.runSynchronously()
asyncFunction("testAsync") { s: String -> s }
```

```js
// App.js
const module = global.ExpoModules.ExpoCellular;
console.log(module.test);
console.log(global.ExpoModules.ExpoCellular == global.ExpoModules.ExpoCellular);
console.log(module.test == module.test);
console.log(module.test("abc"));

console.log(module.testAsync("abc").then((resutl) => console.log(resutl) ))
```

lukmccall committed:
Adjust after rebase

lukmccall committed:
Run spotless

lukmccall committed:
Fix CI

lukmccall committed:
[module-core][Kotlin] register JSI functions using lambdas instead of using a dispatcher (#17501)

# Why

Part of #16977.
To match Swift implementation, we want to register JSI functions using lambdas instead of using a dispatcher. 
Also, it will help in the further when we want to add support for multiple JSI objects.

# How

- Adds a `JNIFunctionBody` and `JNIAsyncFunctionBody` which encapsulates Kotlin's implementation of the exported function and can be recognized by the CPP code.
- Used passed lambdas instead of calling `JavaScriptModuleObject` directly. 

# Test Plan

- bare-expo and simple app:
```kotlin
// Cellular module
function("test") { s: String -> s }.runSynchronously()
asyncFunction("testAsync") { s: String -> s }
```

```js
// App.js
const module = global.ExpoModules.ExpoCellular;
console.log(module.test);
console.log(global.ExpoModules.ExpoCellular == global.ExpoModules.ExpoCellular);
console.log(module.test == module.test);
console.log(module.test("abc"));

console.log(module.testAsync("abc").then((resutl) => console.log(resutl) ))

lukmccall committed:
Fix IllegalArgumentException at RobolectricTestRunner

lukmccall committed:
[skip ci] Add more comments

lukmccall committed:
Add `DoNotStrip` annotation

lukmccall commented on packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp:
```diff
@@ -0,0 +1,136 @@
+      runtime,
+      jsi::PropNameID::forAscii(runtime, metadata.name),
+      metadata.args,
+      [&jsModule = jsModule, cName](
```
Yep, it will. `jsModule` is held by the Kotlin and will live as long as the underlying context.

lukmccall commented on packages/expo-modules-core/android/build.gradle:
```diff
@@ -119,3 +201,124 @@ dependencies {
+// END JNI
+
+// BOOST
+task downloadBoost(dependsOn: createNativeDepsDirectories, type: Download) {
```
boost it needed by the folly, which I'm currently using to make type conversion.

tsapeta commented on packages/expo-modules-core/android/src/main/cpp/JSIInteropModuleRegistry.h:
```diff
@@ -27,12 +30,25 @@ class JSIInteropModuleRegistry : public jni::HybridClass<JSIInteropModuleRegistr
   static void registerNatives();
 
+  /**
+   * Initializes the `ExpoModulesHostObject` and adds it to the globa object.
```
```suggestion
   * Initializes the `ExpoModulesHostObject` and adds it to the global object.
```

tsapeta commented on packages/expo-modules-core/android/src/main/cpp/JSIInteropModuleRegistry.h:
```diff
@@ -27,12 +30,25 @@ class JSIInteropModuleRegistry : public jni::HybridClass<JSIInteropModuleRegistr
+  /**
+   * Initializes the `ExpoModulesHostObject` and adds it to the globa object.
+   *
+   * @param jsRuntimePointer
+   * @param jsInvokerHolder
+   * @param nativeInvokerHolder
```
I don't think we need to list the params if they don't have any description. I'm ok if they don't, it's mostly self-explanatory based on the function's description.

tsapeta commented on packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.h:
```diff
@@ -61,6 +67,17 @@ class JavaScriptModuleObject : public jni::HybridClass<JavaScriptModuleObject> {
+   *  --create JSI function--> `MethodMetadata`
+   *
+   * This abstraction wasn't necessary. However, it makes the management of ownership much easier -
+   * `JavaScriptModuleObject` is hold by the ModuleHolder and `JavaScriptModuleObject::HostObject` is stored in the JS runtime.
```
```suggestion
   * `JavaScriptModuleObject` is held by the ModuleHolder and `JavaScriptModuleObject::HostObject` is stored in the JS runtime.
```

tsapeta commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/jni/JNIFunctionBody.kt:
```diff
@@ -1,12 +1,14 @@
  * This interface is intended to be passed to cpp code.
  * If you want to modify it, please don't forget to change the corresponding jni::JavaClass.
  */
+@DoNotStrip
```
👍

tsapeta reviewed (approved):
🔥 🚀

lukmccall committed:
Apply requested changes

lukmccall committed:
Fix CI

lukmccall committed:
[skip ci] Update packages/expo-modules-core/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>
